daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/transaction/TransactionCoder.scala

```diff
@@ -193,13 +194,19 @@ object TransactionCoder {
                 s"Trying to encode transaction of version $transactionVersion, which requires the exercise return value, but did not get exercise return value in node."))
             case (_, true) => Right(())
           }
-          _ <- (e.key, transactionVersion precedes minContractKeyInExercise) match {
-            case (Some(k), false) =>
+          _ <- (
+            e.key,
+            transactionVersion precedes minContractKeyInExercise,
+            transactionVersion precedes minMaintainersInExercise) match {
+            case (Some(KeyWithMaintainers(_, maintainers)), _, true) if maintainers.nonEmpty =>
+              Left(EncodeError(transactionVersion, isTooOldFor = "maintainers in NodeExercises"))
+            case (Some(KeyWithMaintainers(k, maintainers)), false, _) =>
               encodeVal(k).map { encodedKey =>
                 exBuilder.setContractKey(encodedKey._2)
+                exBuilder.addAllKeyMaintainers(maintainers.toSet[String].asJava)
                 ()
               }
-            case (None, _) | (Some(_), true) =>
+            case (None, _, _) | (Some(_), true, _) =>
               Right(())
```

Comment:
[Cosmetic] 
I found this pattern matching with booleans a bit confusing. 
I woul prefer something like 
```
  _ <- e.key match {
            case Some(KeyWithMaintainers(k, maintainers))
                if !(transactionVersion precedes minContractKeyInExercise) =>
              if ((transactionVersion precedes minMaintainersInExercise) && maintainers.nonEmpty)
                Left(EncodeError(transactionVersion, isTooOldFor = "maintainers in NodeExercises"))
              else
                encodeVal(k).map { encodedKey =>
                  exBuilder.setContractKey(encodedKey._2)
                  exBuilder.addAllKeyMaintainers(maintainers.toSet[String].asJava)
                  ()
                }
            case _ =>
              Right(())
          }
```

Reply:
I also messed up the logic here. I also changed the style now but I'm ambivalent to that.